### PR TITLE
Documentation: Basic internals docs

### PR DESCRIPTION
**Pull request by phyber**
**What this PR does / why we need it**: Adds an `internals` section to the `docs/` directory and adds basic internals documentation for actuators and controllers. Also adds a few `.gitignore`s for `vim` related files.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #505 

**Special notes for your reviewer**: This is by no means complete, but hopefully provides a small base for people to start from.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by randomvariable** [comment]
This has helped us internally get started with Cluster API, hence PRing here.

/lgtm

File: docs/internals/machine_actuator.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# Machine Actuator
+
+The machine actuator is responsible for communicating with the API that will be
```
Comment (davidewatson) [inline docs/internals/machine_actuator.md]: A link to the golang interface would be helpful.

File: docs/internals/machine_actuator.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+`Delete` methods.
+
+A machine in this case is defined as a Linux instance at a cloud provider whose
+parameters are defined within the machine state.
+
+The machine actuator is invoked by the machine controller, which has taken a
+resource request and calls the appropriate machine actuator methods in order to
+realise a machine state.
+
+## Basic Actuator Flow
```
Comment (davidewatson) [inline docs/internals/machine_actuator.md]: This is great!

**Comment by davidewatson** [comment]
Note that, modulo my comments above, I am not necessarily opposed to merging this.

File: docs/internals/controllers.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  - An attempt is made to fetch the appropriate object from Kubernetes; and if
+    that is successful
+  - The object is inspected to see what the appropriate actions to take are
+  - Actuator methods are called to realise the state as necessary
+
+The main difference between the cluster and machine controllers at this level
+is that the machine controller has to take into account that a resource in
+a request may not exist, as such, it must account for things like having to
+create them to begin with.
+The cluster controller has the privilege of knowing that the cluster already
```
[the remark this answers is not in the document]
Comment (phyber) [reply docs/internals/controllers.md]: I was wondering that while writing this. I've been working in the AWS provider, so the above made sense there. I'll try to reword this based on what you've said here.

**Comment by phyber** [comment]
I've realised that the flow I've put under `Basic Actuator Flow` is actually the `Basic Controller Flow`. I'll be taking care of this in some upcoming pushes.

File: docs/internals/controllers.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# Controllers
+
+The main controller implementations ([cluster] and [machine]) are located
+within  the `cluster-api` library, and each perform almost identical basic
+steps within their `Reconcile` methods. The controllers have the responsibility
+of receiving incoming requests and dispatching them to the appropriate actuator
```
Comment (dlipovetsky) [inline docs/internals/controllers.md]: It helps me to think of this in different terms than "a request is received from Kubernetes." Namel,y the controller _watches_ for changes to the resources. For example, the machine controller watches for changes to the machine resource. If a Machine object is created, updated, or deleted, the controller receives that object.

Comment (phyber) [reply docs/internals/controllers.md]: Fair. I've fixed some language in this area to mention the watches, and fixed up the basic flow to mention that after a watch is observed, a reconcile request is received before the controller fetches the cluster/machine object from Kubernetes.

Comment (dlipovetsky) [reply docs/internals/controllers.md]: Thanks a lot!

**Comment by phyber** [comment]
This is ready for another review and merge if there are no more issues.

File: docs/internals/controllers.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+within  the `cluster-api` library, and each perform almost identical basic
+steps within their `Reconcile` methods. The controllers have the responsibility
+of watching for changes to specific resource types and reacting to those
+changes in order to reach a desired state by calling out to the necessary
+actuator methods.
+
+  - Controller watches for changes on a resource type
+  - A change on a watched resource type is observed and the controller
+    receives a reconcile request
+  - An attempt is made to fetch the appropriate object from Kubernetes; and if
```
Comment (roberthbailey) [inline docs/internals/controllers.md]: Is this still true after switching to CRDs? I thought the framework took care of this part now.

Comment (phyber) [reply docs/internals/controllers.md]: This does still appear to be the case in the [machine/controller.go](https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/machine/controller.go) and [cluster/controller.go](https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/cluster/controller.go) files.

**Comment by roberthbailey** [comment]
> Either let's get this PR merged in an acceptable state or get @davidewatson 's gitbook in, but please don't let perfect be the enemy of having some documentation that's helpful to people who are not the original developers of Cluster API.

I totally agree. We discussed during the call today that we want the githbook, but since that will be at least a week out I'm ok merging this in the mean time if you'd prefer.

**Comment by randomvariable** [comment]
From Slack, by @phyber :

> @dwat I hear that the controller documentation from my PR is making its way into the Gitbook in an upcoming PR of yours. That's fine, and I'm happy to close my PR. No point merging it to have it be removed again when the book shows up shortly :slightly_smiling_face: (Will leave the PR open until an ACK later)

**Comment by roberthbailey** [comment]
Closing in favor of https://github.com/kubernetes-sigs/cluster-api/pull/566. 

/close

**Pull request closed without merging**